### DISCUSSION TITLE
How to do selectionset based batching?

### DISCUSSION THREAD

**Discussion by Bjohnson131** [graphql-java/graphql-java#3149] [Q&A] [answered]
Hello, I have types in GraphQL with 2 sorts of fields, fast and slow fields. I'd like to implement batching for my types, but only if the incoming query's selectionset contains slow fields. If the selectionset contains only fast fields, I'd like to use a resolver.

How would one implement this with GraphQL Java?

**Comment by bbakerman** [comment, marked as answer]
See https://www.graphql-java.com/documentation/field-selection

and

https://www.graphql-java.com/blog/deep-dive-data-fetcher-results

I am not sure exactly what you mean by "batching versus resolver" but you can change the way data is fetched by using the selection set

**Reply by Bjohnson131** [reply]
Hey @bbakerman. So If I'm not mistaken, there are two types of resolvers, batch and serial. Serial resolvers seem to be the classic type of resolvers which are shared across other languages in form (DataFetcher), then there are Batch resolvers (BatchLoader/Dataloader), Which are used to collect instances together, and fetch data based on a list of keys.  I'm saying that I want to have both, and use the selectionset for a given query to determine which to use.

**Reply by bbakerman** [reply]
>  I'm saying that I want to have both, and use the selectionset for a given query to determine which to use.

Gotcha - then yes you can look at the selection set in a parent field and decide to use a different strategy of fetching based on it.

> If I'm not mistaken, there are two types of resolvers, batch and serial.

Just to get the nomenclature right - there is not actually two types of resolvers - called DataFecthers in graphql-java.  There is ONLY one type

However they can employ different strategies inside themselves based on selection set

```
DataFetcher<Object> df = env -> {
            Object argInput = env.getArgument("argInput");
            DataFetchingFieldSelectionSet selectionSet = env.getSelectionSet();
            
            if (selectionSet.containsAnyOf("slowField")) {
                DataLoader<Object, Object> dataLoader = env.getDataLoader("slowDataLoaderFetch");
                return dataLoader.load(argInput);
            } else {
                return loadSomeOtherWay(argInput);
            }
        };
```

**Reply by Bjohnson131** [reply]
@bbakerman  This is interesting... Won't calling `dataLoader.load` just result in a bunch of serialized dataloader calls? It won't batch them all into one call, will it?

**Reply by bbakerman** [reply]
> Won't calling dataLoader.load just result in a bunch of serialized dataloader calls?

Have a read of https://www.graphql-java.com/documentation/batching

The data loader mechanism handles the "batching" aspect in its design - it does not "dispatch" the call to get data until it decides that a big enough batch of data is ready.  It does this by tracking what field levels in effect but time is another strategy that could be used.

 
 
>  It won't batch them all into one call, will it?

Yes it will batch if there needs to be batching.  You havent outlined your field structure at all so its hard to be more specific
